background/index.js

```diff
@@ -17,13 +17,171 @@
+    port.onMessage.addListener(processId(port));
+    return;
+  }
+
+  throw new Error(`Unrecognized port: ${port.name}`);
+});
+
+let openedTabId = null;
+let transport = null;
+// Attempt to extract token on start if not available.
+extractToken();
+
+chrome.runtime.onMessage.addListener((request, sender, callback) => {
+  if (!(transport && transport.isConnected) && request.token) {
+    if (transport) {
+      // TODO [ToDr] kill old transport!
+    }
+
+    if (openedTabId) {
+      chrome.tabs.remove(openedTabId);
+      openedTabId = null;
+    }
+
+    console.log('Extracted a token: ', request.token);
+    console.log('Extracted backgroundSeed: ', request.backgroundSeed);
+    chrome.storage.local.set({
+      'authToken': request.token,
+      'backgroundSeed': request.backgroundSeed
+    }, () => {});
+    transport = new Ws(`ws://${UI}`, request.token, true);
+    return;
+  }
+});
+
+let codeCache = null;
+function loadScripts (port) {
+  function retry (msg) {
+    if (msg.type !== 'parity.bar.code') {
+      return;
+    }
+
+    if (!codeCache) {
+      const vendor = fetch(`http://${UI}/vendor.js`)
+        .then(x => x.blob());
+      const embed = fetch(`http://${UI}/embed.html`)
+        .then(x => x.text())
+        .then(page => ({
+          styles: /styles\/embed\.([a-z0-9]{10})\.css/.exec(page),
+          scripts: /embed\.([a-z0-9]{10})\.js/.exec(page)
+        }))
+        .then(res => {
+          return Promise.all([
+            fetch(`http://${UI}/${res.styles[0]}`),
+            fetch(`http://${UI}/${res.scripts[0]}`)
+          ]);
+        })
+        .then(x => Promise.all(x.map(x => x.blob())));
+
+      codeCache = Promise.all([vendor, embed])
+        .then(scripts => {
+          const vendor = scripts[0];
+          const styles = scripts[1][0];
+          const embed = scripts[1][1];
+          // Concat blobs
+          const blob = new Blob([vendor, embed], { type: 'application/javascript' });
+          return {
+            styles: URL.createObjectURL(styles),
+            scripts: URL.createObjectURL(blob)
+          };
+        });
+    }
+
+    codeCache
+      .then(code => port.postMessage(code))
+      .catch(err => {
+        codeCache = null;
+        console.error('Could not load ParityBar scripts. Retrying in a while..', err);
+        setTimeout(() => retry(msg), 5000);
+      });
+  }
 
-  port.onMessage.addListener((message = {}) => {
+  return retry;
+}
+
+function extractToken () {
+  chrome.storage.local.get('authToken', (token) => {
+    if (!token.authToken) {
+      fetch(`http://${UI}`)
+        .then(() => {
+          // Open a UI to extract the token from it
+          chrome.tabs.create({
+            url: `http://${UI}`,
+            active: false
+          }, (tab) => {
+            openedTabId = tab.id;
+          });
+        })
+        .catch(err => {
+          console.error('Node seems down, will re-try', err);
+          setTimeout(() => extractToken(), 1000);
```

Comment:
Might not want a background script making a request every second if I don't have my node up and running on installation. Cf. comment above